collector/uptrack.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"flag"
+	"io/ioutil"
+	"path"
+	"syscall"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	uptrackUpgradePlanPath = flag.String(
+		"collector.uptrack.upgrade_plan_path",
+		uptrackDefaultUpgradePlanPath(),
+		"Path where Uptrack stores its upgrade_plan YAML file.")
+)
+
+type uptrackCollector struct {
+	timeDesc      *prometheus.Desc
+	planCountDesc *prometheus.Desc
+}
+
+func init() {
+	Factories["uptrack"] = NewUptrackCollector
+}
+
+// The location of the upgrade_plan file is based on the currently running kernel.
+func uptrackDefaultUpgradePlanPath() string {
+	var uname syscall.Utsname
+	syscall.Uname(&uname)
+	return path.Join(
+		"/var/cache/uptrack",
```

Comment:
Absolutely no problem. That reasoning makes a lot of sense. At @kumina we've decided that we want to have this collector integrated into the node exporter regardless, as it makes our deployments a lot easier. We'll just keep this collector in our private branch, just like we do already. No time has been wasted on our side. :-)

Given that collectors won't be restricted to just kernel data after all: a colleague of mine, @BartVerc, has written a collector for monitoring Puppet deployments. Would there be any interest in having that upstreamed? Just let me know and I'll ask him to create a pull request.